source/common/common/substitution_format_string.cc

```diff
@@ -0,0 +1,41 @@
+#include "common/common/substitution_format_string.h"
+
+#include "common/access_log/access_log_formatter.h"
+
+namespace Envoy {
+namespace {
+
+absl::flat_hash_map<std::string, std::string>
+convertJsonFormatToMap(ProtobufWkt::Struct json_format) {
+  absl::flat_hash_map<std::string, std::string> output;
+  for (const auto& pair : json_format.fields()) {
+    if (pair.second.kind_case() != ProtobufWkt::Value::kStringValue) {
+      throw EnvoyException("Only string values are supported in the JSON access log format.");
+    }
+    output.emplace(pair.first, pair.second.string_value());
+  }
+  return output;
+}
+
+} // namespace
+
+AccessLog::FormatterPtr SubstitutionFormatStringUtils::fromProtoConfig(
+    const envoy::config::core::v3::SubstitutionFormatString& config) {
+  switch (config.format_case()) {
+  case envoy::config::core::v3::SubstitutionFormatString::FormatCase::kTextFormat:
+    return std::make_unique<AccessLog::FormatterImpl>(config.text_format());
+  case envoy::config::core::v3::SubstitutionFormatString::FormatCase::kJsonFormat: {
+    auto json_format_map = convertJsonFormatToMap(config.json_format());
+    return std::make_unique<AccessLog::JsonFormatterImpl>(json_format_map, false);
+  }
+  case envoy::config::core::v3::SubstitutionFormatString::FormatCase::kTypedJsonFormat: {
+    auto json_format_map = convertJsonFormatToMap(config.typed_json_format());
+    return std::make_unique<AccessLog::JsonFormatterImpl>(json_format_map, true);
+  }
+  default:
+    return nullptr;
```

Comment:
If you add a PGV validation to the oneof to have one of them required to be set, you can switch this to NOT_REACHED and then simplify the error handling to assume a formatter is always returned.

Reply:
In the next [PR](https://github.com/envoyproxy/envoy/pull/11007),  this message is not required.  By returning nullptr to let caller decide the behaviors

Reply:
The message itself is not required, but if the message is used, shouldn't the oneof have a required value? Otherwise the caller can just do what they want if the message is not set at all.

Reply:
Done